docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,112 @@
+title = "SIP 017 - Service Chaining"
+template = "main"
+date = "2024-02-21T23:00:00Z"
+
+---
+
+Summary: Allow in-process HTTP requests to application components without going via the network
+
+Owner(s): ivan.towlson@fermyon.com
+
+Created: Feb 21, 2024
+
+# Requirements and scope
+
+In Spin 2.2, a component can make a HTTP request with only a path (e.g. `/api/accounts`).  Such a request will be routed to the same application, but this is done by Spin prepending the host where the application is running (e.g. `https://mysrsbizness.fermyon.app/api/accounts`).  The consequence of this is that requests go out to the network and come back in, including through load balancers etc.  This adds overhead, and in some environnments could count against quota or incur bandwidth charges.
+
+_Service chaining_ refers to carrying out HTTP requests in memory, without going off the node, by invoking the receiving Spin component directly instead of via the HTTP server.
+
+We have [prototyped](https://github.com/fermyon/spin/pull/2229) dispatching HTTP requests directly to the component handling the route, but this was in the context of an implementation detail of Spin.  The requirement for this project is for an application to declare that it _requires_ dispatch to run locally on the node, and for hosts that cannot meet that prerequisite to reject the application.  This allows developers to write applications as a network of chained microservices, without having to be concerned about what happens if the microservices are not co-located.
+
+# User experience
+
+## Expressing a requirement for chaining
+
+Spin will infer that an application requires service chaining if any component includes a `spin.internal` subdomain (or subdomain wildcard) in its `allowed_outbound_hosts`.
+
+This has two implications:
+
+1. The "allow all" wildcard (`*://*:*`) will not, by itself, enable service chaining.
+2. A templated host (`https://{{ accounts_service }}:*`), with the variable set to a `spin.internal` subdomain, will not, by itself, allow service chaining.
```

Comment:
What is the implication here? It seems like service chaining should still be _permitted_ where a templated host resolves to `spin.internal`, but it just wouldn't trigger the "must support service chaining" app validation?

Reply:
agreed, yes: if we don't have something that we officially treat as a requirement for the semantics of internal service chaining, we should nevertheless still be free to apply the optimization as we see fit.

Reply:
@lann Correct.  So a non-chaining-aware host would fail at runtime while trying to treat `spin.internal` as a routable address, but a chaining-aware one would just keep chaining away like the proverbial boss.